### ISSUE TITLE
SIMD support

### ISSUE THREAD

**Issue by g2p** [BurntSushi/rust-csv#83]
Would SIMD support on Rust nightly be a goal of the project?
I notice that [csvmonkey](https://github.com/dw/csvmonkey) took the lead on the [shootout](https://bitbucket.org/ewanhiggs/csv-game) with SIMD support, so this could make up the difference.

**Comment by BurntSushi** [comment]
Feel free to give it a try. I tried myself and the only thing I learned is that `PCMPESTRI` is incredibly hard to use productively.

This doesn't need a tracking issue. If someone can figure out how to use SIMD to make it even faster, then they should just try it.

**Comment by g2p** [comment]
I have a commit at https://github.com/g2p/rust-csv/commit/simd that uses `x86_mm_cmpestri` (builds on my simd pull requests), unfortunately it is slower. I've looked at `perf annotate` output (after recording with `perf record -b`), and the slowdown seems to be in `csv_core::reader::Reader::read_record` but outside of the added loop.

**Comment by BurntSushi** [comment]
@g2p Yes, that's basically what I tried. Your profile reproduces the same results I found. Basically what happens is that PCMPESTRI is extremely high latency, and it seems this is confirmed by the fact that the `MOV` out of `%ecx` (which contains the substring index) immediately following PCMPESTRI is a hotspot in a profile I took using your code. I found the same thing in my attempt.

Reproducing csvmonkey's performance (on some inputs, IIRC this crate is faster than it on some) probably requires broader changes. I don't think it's worth it.
